spec/hardware/bootdisk.fmf

```diff
@@ -0,0 +1,36 @@
+summary: |
+    Select or create  a guest with a specific boot-disk.
+
+description: |
+    .. code-block::
+       # A list of string, each describing boot disk of beaker guest or root device of openstack.
+       for beaker:
+       boot-disk:
+         - "A Well-Known disk driver/port type"
+
+       for openstack,boot-disk will accept all string params and pass them to root device property.
+       boot-disk:
+         - "device vendor"
+
+       boot-disk:
+         - "device vendor"
+         - "wwn"
```

Comment:
I have an issue with the focus on particular plugins. The specification is supposed to be plugin-agnostic, to hide the implementation detail from the user. It's a declaration that multiple plugins may support, and for each of them one key needs to have the very same meaning. There cannot be a value that has one meaning for Beaker and a different one (or even syntax) for OpenStack (`"A Well-Known disk driver/port type"` vs `"device vendor"`).

This would force the user to target one particular provision plugin, knowing which one it would be in advance, disallowing swapping of provision methods without changing the plan. If I define a plan with OpenStack in mind and run it with `provision --update --how beaker`, what will happen? We cannot tell the user "for `beaker`, you have to rewrite the value into this format". This would be vendor lock-in, just redressed.

Plus, I'm seeing `device vendor` in OpenStack examples, and that I would dare to say should be replaced with a structure - encoding fields into a string makes it harder to manipulate, harder to validate, and harder to consume. We also used these terms in other HW requirements, we already have them well-defined & should re-use them to keep things consistent.

Together, the following would be the specifications I'd propose. It would be more verbose, yes, but it would also provide better visibility of what the user requests, leaving plugins free hands to support the same set of objects. The plugin still may say "we do not support `boot-disk.wwn`", and that is absolutely acceptable - this does not force users to modify their plans when switching plugins. Instead, it calls for other plugins to start supporting this particular HW constraint while preserving its syntax and semantics. So, how about the following?

```yaml
    .. code-block:: yaml

       boot-disk:
           # String, the name of the boot disk driver.
         - driver: nvme

           # String, a network device name.
           device-name: "A Generic Storage Device"

           # Number or string, a storage device ID.
           device: 1234|"> 1234"

           # String, a name of the device vendor.
           vendor-name: "A Well-Known Device Manufacturer"

           # Number or string, an ID of the device vendor.
           vendor: 1234|"> 1234"

           # String, unique storage identifier
           wwn: "0x4000cca77fc4dba1"
```

Reply:
Make sense to me,but it will be a little complicated to implement, the  current code will only accept string array,

for beaker,

```
boot-disk:
   -  pcieport  
```
 
for openstack 

```
boot-disk:
   - "wwn: 0x4000cca77fc4dba"
   -  "vendor:  well-known device manufacturer"
```
And I was thinking to add code to artemis openstack driver to transfer the string to dict and pass them to root-device property.

To implement "The plugin still may say "we do not support boot-disk.wwn", and that is absolutely acceptable "

we need make  wwn/vendor/etc a property of boot-disk for openstack.

```
boot-disk:
    wwn: 0x4000cca77fc4dba
    vendor:  "well-known device manufacturer"
```

Now,I do feel  maybe it's a bad idea to  mix beaker's bootdisk with openstack 's root-device,
ie,maybe we should only handle beaker's bootdisk in this merge request:(

Reply:
> Make sense to me,but it will be a little complicated to implement,

A little complicated + correct is fine. I would be worried if it would be terribly complicated, nearly impossible ;)

> ```
> boot-disk:
>    - "wwn: 0x4000cca77fc4dba"
>    -  "vendor:  well-known device manufacturer"
> ```

Yeah, this is exactly what I want to avoid: this value would be completely useless if I switch from OpenStack plugin (or `artemis`) to `beaker` plugin. The semantics changes based on who's looking at the value, and that's not good. Plus, encoding structure into a string leads to values that are terribly hard to maintain, change, and validate. Hence the structure I proposed above, let's make it explicit. `beaker` driver can always emit a warning when seeing `wwn` key and say "unsupported constraint" - that's acceptable.

> And I was thinking to add code to artemis openstack driver to transfer the string to dict and pass them to root-device property.

As written above, I would fight string-to-dict conversions with all my power :)

> To implement "The plugin still may say "we do not support boot-disk.wwn", and that is absolutely acceptable "
> 
> we need make wwn/vendor/etc a property of boot-disk for openstack.
> 
> ```
> boot-disk:
>     wwn: 0x4000cca77fc4dba
>     vendor:  "well-known device manufacturer"
> ```
> 
> Now,I do feel maybe it's a bad idea to mix beaker's bootdisk with openstack 's root-device, ie,maybe we should only handle beaker's bootdisk in this merge request:(

Good point, maybe we need "root disk" requirement, but I don't think so. It does sound like the same concept: with Beaker, we look for machines with a given "boot disk", with OpenStack (and Artemis with OpenStack pool), we construct a machine with a given "boot disk". We may also look for OpenStack flavors with a given "boot disk" - machines I checked in our OpenStack tenants would be described in Beaker as having `virtio_blk` bootdisk key. If I put `driver: virtio_blk` into my plan, I might get an OpenStack VM, based on the existing flavor, as well as a KVM VM from Beaker. So from my PoV, "constructing a VM with given boot/root disk" fits the `boot-disk` description too. I'd say this aspect is fine.

The issues lie in the specification, I'm afraid, with string-only items of the `boot-disk` array and encoding of details into those strings. That harms the possibility of swapping provision plugins and makes it harder to work with the resulting plan. I believe we can do better, even if the cost is a slightly more verbose specification and slightly more complicated implementation.

And I don't think the implementation will be that complicated, on the other hand, I'm pretty sure it will be cleaner than working with `"vendor:  well-known device manufacturer"`. Let's ignore the implementation in various drivers for now, and focus on clean, correct, and easy-to-understand specifications. In my experience, it often contributes heavily to getting a cleaner code eventually.